massa-execution/src/worker.rs

```diff
@@ -240,121 +239,194 @@ impl ExecutionWorker {
             self.cfg.clock_compensation,
         )?;
         if let Some(end_step) = end_step {
-            while self.last_active_slot < end_step {
-                self.last_active_slot =
-                    self.last_active_slot.get_next_slot(self.cfg.thread_count)?;
-                self.push_request(ExecutionRequest::RunFinalStep(ExecutionStep {
+            // slot S
+            let s = self.last_active_slot.get_next_slot(self.cfg.thread_count)?;
+
+            while s <= end_step {
+                // call the VM to execute an SCE-active miss at slot S
+                self.push_request(ExecutionRequest::RunActiveStep(ExecutionStep {
                     slot: self.last_active_slot,
                     block: None,
                 }));
+
+                // set last_active_slot = S
+                self.last_active_slot = s;
+
+                s = s.get_next_slot(self.cfg.thread_count)?;
             }
         }
         Ok(())
     }
 
+    /// checks whether a miss at slot S would be SCE-final by looking up subsequent CSS-final blocks in the same thread
+    /// see spec at https://github.com/massalabs/massa/wiki/vm-block-feed
+    ///
+    /// # Arguments
+    /// * s: missed slot
+    /// * max_css_final_slot: maximum lookup slot (included)
+    fn is_miss_sce_final(&self, s: Slot, max_css_final_slot: Slot) -> bool {
+        let mut check_slot = Slot::new(s.period + 1, s.thread);
+        while check_slot <= max_css_final_slot {
+            if self.pending_css_final_blocks.contains_key(&check_slot) {
+                break;
+            }
+            check_slot.period += 1;
+        }
+        return check_slot <= max_css_final_slot;
+    }
+
     /// called when the blockclique changes
+    /// see spec at https://github.com/massalabs/massa/wiki/vm-block-feed
     fn blockclique_changed(
         &mut self,
-        mut blockclique: BlockHashMap<Block>,
-        mut finalized_blocks: BlockHashMap<Block>,
+        blockclique: BlockHashMap<Block>,
+        finalized_blocks: BlockHashMap<Block>,
     ) -> Result<(), ExecutionError> {
-        // filter out any already-finalized blocks
-        // this can happen for example on bootstrap because the consensus vs execution bootstrap snapshots may be out of sync
-        blockclique.retain(|_b_id, b| b.header.content.slot > self.last_final_slot);
-        finalized_blocks.retain(|_b_id, b| b.header.content.slot > self.last_final_slot);
+        // 1 - reset the SCE state back to its latest final state
 
-        // stop the current VM execution and reset state to final
+        // revert the VM to its latest SCE-final state by clearing its active slot history.
         // TODO make something more iterative/conservative in the future to reuse unaffected executions
         self.reset_to_final();
 
-        // stop the current VM execution and reset state to final
-        // TODO reset VM
+        // set `last_active_slot = last_final_slot
```

Comment:
Not sure about that comment

Reply:
it's just there so that every comment matches word-to-word one line in the spec, and that both can be read linearly

Reply:
maybe here it's a little bit overkill to just repeat the code in a comment.

Reply:
removed